### PR TITLE
feat!: add new Discovery Web IDL definitions

### PR DESCRIPTION
**Pull request by JKRhb**
This is an attempt to create Web IDL definitions based on the example in https://github.com/w3c/wot-scripting-api/issues/364#issuecomment-1137703749. This is still a work in progress and far from mergeable, but we might already be able to use for discussing the details of the new API.

The most important changes are the addition of the [async iterable](https://webidl.spec.whatwg.org/#idl-async-iterable) keywords and the removal of the `next()` method (which is now available on the inherited `Symbol.asyncIterator` field). Assuming that the discovery process should start right away, I also removed the `start()` method. However, if it is good practice to have such a method in a web API, then it could of course still be included. 

Furthermore, as both the `direct` and the `directory` methods need URLs in order to work, I made the `filter` and its `url` field required. This could be either reverted if other discovery mechanisms should be supported, or sub-interfaces for the these two methods could be introduced, where `url` is in fact a required field.

***
<a href="https://pr-preview.s3.amazonaws.com/JKRhb/wot-scripting-api/pull/405.html" title="Last updated on Oct 10, 2022, 11:41 AM UTC (099e1e6)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/wot-scripting-api/405/e94e160...JKRhb:099e1e6.html" title="Last updated on Oct 10, 2022, 11:41 AM UTC (099e1e6)">Diff</a>

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -3830,7 +3832,8 @@ <h3>The <dfn>start()</dfn> method</h3>
             Whenever a new <a>Thing Description</a> |td:ThingDescription| is discovered by the underlying platform, run the following sub-steps:
             <ol>
               <li>
-                <a href="https://fetch.spec.whatwg.org/#fetch-api">Fetch</a> |td| as a JSON object |json|. If that fails, set the <a href="#dom-thingdiscovery-error">error</a> property to {{SyntaxError}}, discard |td| and continue the discovery process.
+                Retrieve |td| as a JSON object using the <a>Protocol Binding</a> specified by the |filter|'s <dfn>url</dfn> property.
```
Comment (zolkis) [inline index.html:3835]: I would add a note that previously this was using Fetch, which remains a possibility, depending on binding.

File: index.html [hunk 2]
```diff
@@ -3879,13 +3882,10 @@ <h3>The <dfn>next()</dfn> method</h3>
...
           <li>
-            If the <a href="#dom-thingdiscovery-active">active</a> property is `true`, wait until the <a>discovery results</a> <a>internal slot</a> is not empty.
-          </li>
-          <li>
-            If <a>discovery results</a> is empty and the <a href="#dom-thingdiscovery-active">active</a> property is `false`, set the <a href="#dom-thingdiscovery-done">done</a> property to `true` and reject |promise|.
+            If the <a href="#dom-thingdiscovery-active">active</a> property is `true`, call the `next` method on the Discovery object's Symbol.AsyncIterator.
           </li>
           <li>
-            Remove the first {{ThingDescription}} object |td| from <a>discovery results</a>.
+            If the `value` property in the yielded result from the Symbol.AsyncIterator is undefined and the `done` property is set to `true`, set the <a href="#dom-thingdiscovery-done">done</a> property to `true` and reject |promise|.
```
Comment (zolkis) [inline index.html:3888]: We need at least one example for how to handle async iterators in client code.

File: index.html (hunk 1 above)
Comment (JKRhb) [reply index.html]: I added the line in 704191d2fa0925d3bbe632c57131cfd05583c934 for now, let me know if you see potential for improvement :)

File: index.html (hunk 2 above)
Comment (JKRhb) [reply index.html]: I adjusted the first discovery example in 9646fb627c83b257cc671e70d4938ed0767d932d, I hope this points into the right direction.

**Comment by JKRhb** [comment]
Made a couple of additional adjustments regarding the use of the `AsyncIterator`, removing mentions of the internal `results` slot. I think there is still some work to be done regarding the prose and the second example, but I suppose the current state can be used as a starting point for further discussion :) Therefore, I'd remove the draft status of the PR, indicating that it is ready for further reviews. (Thank you, @zolkis, for the feedback you already gave above!)

File: index.html (hunk 1 above)
Comment (zolkis) [reply index.html]: Looks good!

File: index.html (hunk 2 above)
Comment (zolkis) [reply index.html]: Fair.

File: index.html [hunk 3]
```diff
@@ -3761,8 +3762,8 @@ <h3>The <dfn>ThingFilter</dfn> dictionary</h3>
       </p>
       <pre class="idl">
         dictionary ThingFilter {
-          DiscoveryMethod method = "directory";
-          USVString? url;
+          DiscoveryMethod method = "direct";
+          required USVString url;
```
Comment (zolkis) [inline index.html:3766]: About this, let's discuss in the call.

Comment (danielpeintner) [reply index.html:3766]: Scripting call 2022-09-26
* Move out this change in another PR, keep this PR a syntactic change 
* `auto` method ?

Comment (JKRhb) [reply index.html]: Moved these lines to a new PR #434 now :)

**Review by zolkis — APPROVED** [review]

**Comment by danielpeintner** [comment]
Scripting Call 2022-10-10
-> Changes Approved

**Pull request merged**